# [Model] Data Analytics Test - iFood
###### [By Israel Mendes](https://github.com/israelmendez232)

This notebook were generated locally, because of the complexity to run the models that won't work for the [Databricks](https://databricks.com/) Community Edition.

The main library used here is [PyCaret](https://pycaret.readthedocs.io/en/latest/) to generate the model and tuning, because is very pratical and productive. The summary:
1. Train the model and explorations;
2. Validate the results and save the model;
3. Avaluate the return;
4. [EXTRA] Results

## 1. Train the model
Bring more explorations.

In [1]:
import pandas as pd
from pycaret.classification import *

df = pd.read_csv("../data/data_analytics_cleaned.csv") 

df.head(10)

ID  Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
0  5524        1957  Graduation         Single  58138.0        0         0   
1  2174        1954  Graduation         Single  46344.0        1         1   
2  4141        1965  Graduation       Together  71613.0        0         0   
3  6182        1984  Graduation       Together  26646.0        1         0   
4  5324        1981         PhD        Married  58293.0        1         0   
5  7446        1967      Master       Together  62513.0        0         1   
6   965        1971  Graduation       Divorced  55635.0        0         1   
7  6177        1985         PhD        Married  33454.0        1         0   
8  4855        1974         PhD       Together  30351.0        1         0   
9  5899        1950         PhD       Together   5648.0        1         1   

  Dt_Customer  Recency  MntWines  MntFruits  MntMeatProducts  MntFishProducts  \
0  2012-09-04       58       635         88              546              172   
1  2014-03-08       38        11          1                6                2   
2  2013-08-21       26       426         49              127              111   
3  2014-02-10       26        11          4               20               10   
4  2014-01-19       94       173         43              118               46   
5  2013-09-09       16       520         42               98                0   
6  2012-11-13       34       235         65              164               50   
7  2013-05-08       32        76         10               56                3   
8  2013-06-06       19        14          0               24                3   
9  2014-03-13       68        28          0                6                1   

   MntSweetProducts  MntGoldProds  NumDealsPurchases  NumWebPurchases  \
0                88            88                  3                8   
1                 1             6                  2                1   
2                21            42                  1                8   
3                 3             5                  2                2   
4                27            15                  5                5   
5                42            14                  2                6   
6                49            27                  4                7   
7                 1            23                  2                4   
8                 3             2                  1                3   
9                 1            13                  1                1   

   NumCatalogPurchases  NumStorePurchases  NumWebVisitsMonth  AcceptedCmp3  \
0                   10                  4                  7             0   
1                    1                  2                  5             0   
2                    2                 10                  4             0   
3                    0                  4                  6             0   
4                    3                  6                  5             0   
5                    4                 10                  6             0   
6                    3                  7                  6             0   
7                    0                  4                  8             0   
8                    0                  2                  9             0   
9                    0                  0                 20             1   

   AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  Complain  \
0             0             0             0             0         0   
1             0             0             0             0         0   
2             0             0             0             0         0   
3             0             0             0             0         0   
4             0             0             0             0         0   
5             0             0             0             0         0   
6             0             0             0             0         0   
7             0        

In [2]:
# Selecting only the best variables estipulated on the 02-EDA phase.
df_focused = df[["ID", "AcceptedCmp5", "AcceptedCmp1", "MntWines", "AcceptedCmp3", "MntMeatProducts", "NumCatalogPurchases", "Recency", "Teenhome", "Kidhome", "Response"]]
df_focused.head()

ID  AcceptedCmp5  AcceptedCmp1  MntWines  AcceptedCmp3  MntMeatProducts  \
0  5524             0             0       635             0              546   
1  2174             0             0        11             0                6   
2  4141             0             0       426             0              127   
3  6182             0             0        11             0               20   
4  5324             0             0       173             0              118   

   NumCatalogPurchases  Recency  Teenhome  Kidhome  Response  
0                   10       58         0        0         1  
1                    1       38         1        1         0  
2                    2       26         0        0         0  
3                    0       26         0        1         0  
4                    3       94         0        1         0

In [3]:
df_focused.describe()

ID  AcceptedCmp5  AcceptedCmp1     MntWines  AcceptedCmp3  \
count   2213.000000   2213.000000   2213.000000  2213.000000   2213.000000   
mean    5586.899232      0.072752      0.064166   305.153638      0.073656   
std     3247.819194      0.259788      0.245104   337.305490      0.261269   
min        0.000000      0.000000      0.000000     0.000000      0.000000   
25%     2815.000000      0.000000      0.000000    24.000000      0.000000   
50%     5455.000000      0.000000      0.000000   175.000000      0.000000   
75%     8420.000000      0.000000      0.000000   505.000000      0.000000   
max    11191.000000      1.000000      1.000000  1493.000000      1.000000   

       MntMeatProducts  NumCatalogPurchases      Recency     Teenhome  \
count      2213.000000          2213.000000  2213.000000  2213.000000   
mean        166.962494             2.671487    49.007682     0.505648   
std         224.226178             2.927096    28.941864     0.544236   
min           0.000000             0.000000     0.000000     0.000000   
25%          16.000000             0.000000    24.000000     0.000000   
50%          68.000000             2.000000    49.000000     0.000000   
75%         232.000000             4.000000    74.000000     1.000000   
max        1725.000000            28.000000    99.000000     2.000000   

           Kidhome     Response  
count  2213.000000  2213.000000  
mean      0.441934     0.150474  
std       0.536965     0.357617  
min       0.000000     0.000000  
25%       0.000000     0.000000  
50%       0.000000     0.000000  
75%       1.000000     0.000000  
max       2.000000     1.000000

In [4]:
# Defining the rows for testing and prediction
df_train = df_focused.sample(frac = 0.9, random_state = 786)
df_prediction = df_focused.drop(df_train.index)

df_train.reset_index(drop = True, inplace = True)
df_prediction.reset_index(drop = True, inplace = True)

print('Data for Training/Modeling: ' + str(df_train.shape))
print('Data For Predictions: ' + str(df_prediction.shape))

Data for Training/Modeling: (1992, 11)
Data For Predictions: (221, 11)


In [5]:
# Setup for the classification model
setup_sample = setup(df_train, target = 'Response')

In [6]:
# Compare different models
best_model = compare_models()

The best models are highlighted in yellow. Decided to move with CatBoost + Extra Trees and Naive Bayes.

In [7]:
catboost = create_model('catboost')

In [8]:
et = create_model('et')

In [9]:
nb = create_model('nb')

In [10]:
# Using Hyperparameters
tuned_catboost = tune_model(catboost)

In [11]:
# Using Hyperparameters
tuned_et = tune_model(et)

In [12]:
# Using Hyperparameters
tuned_nb = tune_model(nb)

In [13]:
# Blending the models
blender = blend_models(estimator_list = [tuned_catboost, tuned_et, tuned_nb], method = 'soft')

In [ ]:
stacker = stack_models(estimator_list = [tuned_catboost, tuned_et, tuned_nb], meta_model = catboost)

IntProgress(value=0, description='Processing: ', max=6)

Initiated  . . . . . . . . . . . . . . . . . .             13:11:45
Status     . . . . . . . . . . . . . . . . . .     Fitting 10 Folds
Estimator  . . . . . . . . . . . . . . . . . .  Stacking Classifier

Empty DataFrame
Columns: [Accuracy, AUC, Recall, Prec., F1, Kappa, MCC]
Index: []

Chosing stacker over blender because of a significant better performance on stacker based on Recall, Kappa and MCC.

In [ ]:
plot_model(stacker)

In [ ]:
plot_model(stacker, plot = 'confusion_matrix')

In [ ]:
plot_model(stacker, plot = 'boundary')

## 2. Validate the results
Save the model

In [ ]:
predict_new = predict_model(stacker, data = df_prediction)
predict_new.tail()

In [ ]:
predict_new['result'] = predict_new['Response'] == predict_new['Label']
predict_new.tail()

In [ ]:
lost = predict_new[predict_new['result'] == False]['result'].count()
won = predict_new[predict_new['result'] == True]['result'].count()
end_result = won / (won + lost)

print(f"Right predictions: {won}")
print(f"Wrong predictions: {lost}")
print(f"End result in % of the predictions: {end_result}")

In [ ]:
# Saving locally the model
save_model(stacker, model_name='end-classifier-model')

## 3. Avaluate the return
Now it's time to validate if the insights are accurate based on the **prediction sample**. Main points to avaluate:
 - Prove that the insight could have a **campaign rate beyond 15%**, which was the standart for this campaign;
 - Considering the 6.720MU spended on this "sample", with 2.213 (not counting the null and outliers). This campaign has **invested 3.03MU/customer**;
 - The **total received** by each Response was: 3.674MU / 333 (not counting the null and outliers) => **11.03MU/customer**;
 - To be sucessful, the **campaign rate needs to be 28% or higher** (3.03 / 11.03) based on the cost and return over customer;
 - The **R.O.I.** was (3.674MU / 6.720MU) - 1 => **-45,32%**;
 
Since we can't change the price or the campaign, we can focus more on segmentation and the historical data of those customers. And to provide better insights to overcame the standart elements mentioned earlier.

In [ ]:
# Considering we will only impactate the customers that were predicted as 1, avoing the 0 ones.
validation = predict_new.copy()
validation = validation[validation['Label'] == 1]
validation.head()

In [ ]:
base_total_customers = predict_new['result'].count()
customers_impactated = validation['result'].count()
rel_customers_imp_total_base = customers_impactated / base_total_customers
customers_positive = validation[predict_new['result'] == True]['result'].count()


total_spend = customers_impactated * 3.03 
total_return = customers_positive * 11.03

campaign_rate = customers_positive / customers_impactated
roi = (total_return / total_spend) - 1

print(f"Total customers impactated: {customers_impactated}")
print(f"Customers impactated and positive response: {customers_positive}")
print(f"Customers impactated / total customers: {rel_customers_imp_total_base}")
print(f"Total spend on campaign: {total_spend}")
print(f"Total return on the campaign: {total_return}")
print(f"Campaign rate: {campaign_rate}")
print(f"ROI: {roi}")

## [EXTRA] Results

The model was responsable to:
- Increase the **Campaign Rate** to 15% up to **92,30%**;
- Deliver a **ROI** to **+236.02%**, compared to -45,32% of the starndart campaign.

---

If we extrapolate the results from this model, it would generate the following results in the whole base:
- 2,213 (whole base) * 0.058 (rate from impactated customers) * 3.03 (cost by customer) => **388MU in costs;**
- 2,213 (whole base) * 0.058 (rate from impactated customers) * 0.923 (campaign rate) * 11.03 (return by customer) => **1.306MU in profit;**
- The model would save up to **6.332MU**;
- Also, the model would bring 35.54% of the standart results of the campaign with ONLY 5.77% of the total budget!